### PR TITLE
Split QPDFObject into QPDFObject and QPDFValue

### PR DESCRIPTION
**Pull request by m-holger**
Early draft for feedback

### CODE REVIEW

File: libqpdf/QPDFObjectHandle.cc
Comment (jberkenbilt) [inline libqpdf/QPDFObjectHandle.cc:3133]: By removing this, you are causing the code to have to hang unto a bunch of reserved objects that are going to get replaced later. It's possible that the original code was made functionally unnecessary by the checks for whether the object was changed (a relatively recent addition), but I wonder if there is a performance impact. It's possible that the old code is superfluous.

File: libqpdf/QPDF.cc
Comment (jberkenbilt) [inline libqpdf/QPDF.cc]: stray ;

**Comment by jberkenbilt** [comment]
It seems to me that this pull request is doing multiple things. Some of it I think is suitable to accept sooner. For example, if the change to how reserved objects work is a good change without a negative performance impact, it may be possible to merge that change on its own.

File: include/qpdf/Constants.h
Comment (m-holger) [reply include/qpdf/Constants.h:85]: `ot_unresolved` should go to the first section with `ot_uninitialized` and `ot_reserved`. However, this effects a fair bit of test output. In view of that, I postponed putting it into its proper place until it is clearer whether this PR is going to go ahead or not.

File: include/qpdf/QPDFObject.hh
Comment (m-holger) [reply include/qpdf/QPDFObject.hh:54]: If this goes ahead Constants.h will be changed  to the same order.

**Comment by jberkenbilt** [comment]
I think I am okay with the general direction here. I am not at this time comfortable with how unresolved or reserved objects are handled. I haven't fully wrapped my head around the implementation. I think if we can get rid of `objectChanged`, it will be a win.

I know you want this in qpdf 11, but I don't think I am comfortable with this going into qpdf 11. qpdf 11 has a lot of changes in it, and I'd like to get it released. I'm going to need some time to get comfortable with this and to convince myself that the lines are drawn in the right place.

I'm still not convinced of the need for separate `QPDFObject` and `QPDFValue` types, but maybe I'm just not seeing what problem this is solving that can't be solved without splitting the type. I need to dig a little deeper. I see the problem of wanting there to be only one copy in existence of the underlying object with a specific og within a QPDF. I know we've been through this, but I guess I'm still not completely seeing it. Can we take a step back and make sure I fully understand the problem you're trying to solve? Then it will be easier for me to decide whether I agree with this approach to solving it.

File: include/qpdf/Constants.h
Comment (jberkenbilt) [reply include/qpdf/Constants.h:85]: I don't think we should ever change what numerical values are assigned to any of the existing ot_ values. We'll have to live with it not being where you think it should go.

File: include/qpdf/QPDFObject.hh
Comment (jberkenbilt) [reply include/qpdf/QPDFObject.hh:54]: No, I don't think we can change the order in Constants.h

File: libqpdf/QPDFObjectHandle.cc
Comment (m-holger) [reply libqpdf/QPDFObjectHandle.cc:3162]: I am happy with either "this->" or no "this->". I just like a clear rule. That said, I find "this->m->" pointless as I don't think that there can ever be any confusion

**Comment by m-holger** [comment]
Not sure how I started a review

File: libqpdf/QPDFObjectHandle.cc
Comment (jberkenbilt) [reply libqpdf/QPDFObjectHandle.cc:3162]: Life is messy, and there are not always clear rules. I have no problem without `m` vs. `this->m`. As I am unlikely to accept a pull request that goes through and systematically makes all use of `this->` consistent, you may have to just accept the inconsistency. I used to get hung up about this stuff, but over the years, I've just accepted that people's brains are not consistent over time. Styles change, best practices change, and preferences changes. Perhaps where you don't like inconsistency, I don't like churn. Every unnecessary change carries risk and cognitive load. Maybe it's different for you, but for me, the cognitive load of dealing with the inconsistency is very low compared to the cognitive load of dealing with lots of small, tactical changes to the code base for reasons we've discussed plenty of times already.

**Comment by m-holger** [comment]
> It seems to me that this pull request is doing multiple things. Some of it I think is suitable to accept sooner. For example, if the change to how reserved objects work is a good change without a negative performance impact, it may be possible to merge that change on its own.

The commits at this point are geared towards demonstrating that splitting `QPDFObject` allows `QPDF::objectChanged` to be removed at an acceptable cost. I believe this has been achieved.

At the time of pushing the branch, the combined effect of the changes was an improvement in runtime of somewhere between 1.5% and 2%.

The removal of  `QPDF::objectChanged` and the related `m::ever_replaced_objects` is implemented in commit "Remove QPDF::ObjectChanged". The removal is made possible by mutating `obj_cache[og].object` in place (rather than replacing the existing object in the cache) in `QPDF::replaceObject` and `QPDF::swapObjects`. 

The ability to mutate objects in place relies on the methods `QPDFObject::assign` and `QPDFObject::swapWith` introduced in commit "Split QPDFObject into QPDFObject and QPDFValue". These two methods in turn depend on the pdf object type being able to be changed by replacing the  `QPDFObject::value` member rather than the entire `QPDFObject` itself.

The ability to mutate objects associated with a pdf  file transparently without the need to refresh any object handles also relies on all object handles associated with an `QPDF` object  to have a valid `QPDFObject` in their `obj` member. Commits "Add class QPDF_Unresolved" and "Modify newIndirect to set QPDFObjectHandle::obj" are in furtherance of this need, and commits "Modify newIndirect to set QPDFObjectHandle::obj" and "Remove QPDFObjectHandle::makeReserved" are associated clean-up commits.
[From a simplicity perspective I considered to also implement uninitialized `QPDFObjectHandle`s as  `QPDFObject`s rather than a bool flag, removing the need to ever check  `QPDFObjectHandle::obj` for being a nullptr. However, this turned out to be ridiculously expensive.]

Commits "Replace calls to QPDFObjectHandle::Factory::newIndirect where possible" and "Add private methods QPDFObjectHandle::asArray etc" primary purpose is to facilitate the implementation of "Split QPDFObject into QPDFObject and QPDFValue". 

The purpose of "Change QPDFObjectHandle::shallowCopyInternal to copy scalars" is to ensure that no object can appear more than once in the   `QPDF`  `obj_cache`.

The (possibly secondary) purpose of commits "Add private methods QPDFObjectHandle::asArray etc",  "Change QPDFObjectHandle::shallowCopyInternal to copy scalars" and "Remove virtual methods QPDFValue::getTypeCode and getTypeName" is to mitigate the cost of the extra indirection involved in the split of `QPDFObject`.

**Comment by jberkenbilt** [comment]
Excellent clarification. What I'd like to do, and I will do this before making a final decision about qpdf 11, is to walk through this commit by commit in my own branch and play with some alternatives of the things that feel uncomfortable to me. This will either lead me to agree with your conclusions (having convinced myself that my objections are unfounded) or suggest an alternative. I also want to think through whether there should be another case added to the performance test. I think the whole reserved object thing is most likely going to be most heavily exercised by operations involving moving pages around between files. There are some very subtle cases around copied streams, particularly when a stream with a stream data provider is copied from one file to another. There is no way to directly exercise that from the command line, but I have some cases in test_driver for it.

I may also accept some commits as they are by cherry-picking them. I don't think you care that much, but I will retain your authorship of commits that are predominantly or entirely yours. I'm not sure whether I will be able to get to it this weekend. Then I'm on vacation the week of August 13. I have a green light from work to do a focused qpdf block after that as I am winding down a complex project I've been working on for a few months. So before much longer, I will be able to fully devote my attention to this.

In spite of my constant push back, I consider this to be great work, and I appreciate your willingness and ability to deal with my quirks. I think this work will likely resolve one of the thorniest issues about the qpdf internals, so it's worth the time.

Resolving the conflicts here would be helpful but not required. I'll probably merge #746, so you may want to wait for that in case it conflicts with this.

**Comment by m-holger** [comment]
> I know you want this in qpdf 11

Actually, this is not a priority for me.

My next step is to move `QPDFObjectHandle::qpdf` and `og` to `QPDFValue`. It might be nice to get that one in as it reduce the footprint to a max of 16 bytes. (qpdf is done, but needs tidying. og needs more work).

[There is a question in my mind over #738. If changing the implementation of the inline functions breaks the ABI and the move doesn't happen than the inlining probably should be restricted to internal use, at least for the time being. That is where the performance gain is anyway.]

Other than the above, I prefer not to be pushed by the qpdf11 deadline. I feel the changes I am suggesting are complex enough to warrant unpressured consideration.

**Comment by m-holger** [comment]
> is to walk through this commit by commit in my own branch

Excellent idea. I've been feeling that I am working a bit in a vacuum on this. A bit of detailed challenge would be good to develop my own thinking.

**Comment by jberkenbilt** [comment]
Good point about inlining and ABI. Inlining of course removes them from the ABI. (And to be sure, I just checked with `check_abi dump`. For qpdf 11, obviously ABI changes are okay since this is a major release, but incremental updates between here and qpdf 12 won't be allowed to break ABI. Can you think of any reason why breaking the ABI in this way is bad for qpdf 11? I'm not that concerned about this for Windows DLL users who load functions at runtime since they are using the C API rather than C++ API, so I think it's okay if these disappear from the ABI.

The main thing I would like to make sure we get in before qpdf 11 is if there are ABI-breaking changes that will make it easier for us to make changes incrementally without breaking ABI again before qpdf 12, but really it's not that big a deal. There's no rule that says qpdf 12 can't follow close behind qpdf 11.

I would still like to do a deep study of your changes so that I can provide more meaningful input on the implementation. Either I will get fully behind it or have a concrete suggestion as I said. I'd like to do this before I close the window on API/ABI changes for qpdf 11. The next major feature I was planning on working on is an enhanced API for dealing with page-level operations. I anticipate that I will be able to do this without ABI change and so these will likely be minor qpdf 11 releases. We can have the object/value split, or whatever it ends up being, as the main epic for 12. I probably have no more than one to two days' work left otherwise before I am ready to go on qpdf 11.

So my plan then I guess will be to wrap up changes for qpdf 11 over the next two to three weeks with a surge in late August/early September if I can't get it done on weekends. This will include a full analysis of your changes and possibly cherry-picking some of the earlier commits that appear to be advantageous regardless of whether/how we do this.

**Comment by m-holger** [comment]
> I also want to think through whether there should be another case added to the performance test.

I think performance tests probably deserve some consideration.

I certainly found many surprises when I ran the tests. It is probably worth thinking about making performance testing more of a part of the normal work-flow.

A would be nice if the tests could be run a bit more quickly. A couple of ideas -
- have a separate build directory for each commit id
- don't delete the temp directory
- don't rebuild if the directory for the commit id already exists.
- give option to skip api check; only build what's required for performance tests.
- codify the data used for performance tests.

That said, I am not keen to touch either bash or perl scripts myself.

**Comment by jberkenbilt** [comment]
Can you give me a quick rundown of how you are currently using performance testing? Maybe I can come up with some improvements. I tend to run performance tests fairly infrequently in my normal work and sometimes cache results. I'm careful to always run them in the same way and with my machine otherwise idle so I can get meaningful comparisons. A big part of the slowness is that things are run over multiple iterations to get more statistically meaningful results. I might be able to come up with a smoke test approach (if there isn't already one there) that is less statistically accurate but still provides meaningful benchmarks for higher impact changes. I'll give it some thought as I'm sure I'll be running them a lot more when I play with the changes.

**Comment by m-holger** [comment]
> Good point about inlining and ABI. Inlining of course removes them from the ABI. (And to be sure, I just checked with `check_abi dump`

Sorry, I am not sure I understand this. Can we change the implementation without breaking the ABI?

I am not that clear on the ABI. Just to confirm my understanding, can you confirm that the following do not break the ABI:
- adding new methods(including .overloads to existing methods) or classes to the public API
- changing the private API

**Comment by jberkenbilt** [comment]
> Sorry, I am not sure I understand this. Can we change the implementation without breaking the ABI?

Lots of things can be changed without breaking ABI. 

> I am not that clear on the ABI. Just to confirm my understanding, can you confirm that the following do not break the ABI:
> 
> * adding new methods(including .overloads to existing methods) or classes to the public API
> * changing the private API

Generally, these both preserve ABI. However, changing something to inline does break the ABI because it no longer appears as a symbol in the shared library. Changing something from inline doesn't break the ABI but it adds to the ABI and requires a minor version bump.

In C++, there are a few gotchas. Changing the size/shape of anything that is ever stack allocated breaks ABI, which is why I use the shared_ptr<Members> pattern all over the place. Mostly I disallow copying of that so I'm not creating reference semantics (with QPDFObjectHandle and JSON being notable exceptions) and, for non-critical sections, I'm willing to pay the small performance hit in favor of the flexibility of adding member variables. Also adding a virtual function to a class that didn't previously have virtual functions breaks ABI.

My check_abi script does a decent but not perfect job of detecting ABI changes. I used to rely on running the test suite on an older version of the binaries with a newer version of the shared library, but when I introduced QPDFJob, this became unreliable because the qpdf CLI now hardly exercises any public API, and when QPDFJob calls it, it is not traversing the shared object barrier, which would make it possible to not notice ABI breakage. Also, the code in libtests links with an object library which makes it possible to unit test private interfaces but also removes the shared object barrier. Bottom line: my old method for detecting ABI breakage doesn't work, hence check_abi.

There are subtle ways to break ABI as well, like changing values of constants. Even if the calling interface is the same, changing the meaning of some value passed to a function breaks ABI. This is why I don't want to change the values of the ot_ types in Constants.h. While a recompile would solve it for C++ users, consumers of the ABI through shared loading across languages would not get any trigger, and anyone relying on object types through that method would not have any automated way to update their mapping between constants and their values. As such, trying to write code that worked properly across changes to constant values would require having the application code know what values the constants would have across versions that are only knowable at runtime. That would be a disaster for the programmer, which is why I won't change the values of constants without an extremely compelling reason. In fact, if I were to remove an object type, I would leave something like `unused1` in its place to avoid changing the constant values.

**Comment by m-holger** [comment]
> Can you give me a quick rundown of how you are currently using performance testing?
Basically I run a slightly doctored version of abi-perf-check (with code checking for ABI changes commented out and slightly doctored output) as a batch job with typically 5 runs per comparison.

I find there is a reasonable variability even with my browser closed and nothing else happening on my machine

**Comment by jberkenbilt** [comment]
So I guess ABI breakages that require recompilation/linking but don't require source changes are allowed freely at major version releases, so things like collapsing overloaded methods with single methods that have default values, changing what's inlined, changing constructors/destructors from whether they use `= default` or not, etc. are all open at major version boundaries. Things that are not detectable across language barriers, like changing constant values, are never allowed unless all code paths that use the old constants are removed. API updates are allowed but only sparingly and only with an easy migration path.

**Comment by jberkenbilt** [comment]
I'll be away from my computer for a few hours -- possibly not back before the end of the day in the UK.

I usually run performance_check directly. I only use api-perf-check at release time. I think you will find that running performance_check directly should help with most of your issues. You can also vary time and iterations there already.

**Comment by jberkenbilt** [comment]
It dawned on me that the files I use for performance testing are not published anywhere. I will have to do something about that. What an odd oversight. I guess I must be the only one who runs it. I'm not sure how you're even able to get useful information, though I guess you must have supplied your own files. I also see that my json test file is broken, but no one would notice that since the test file is not published anywhere. One of my test files is not redistributable. I use the first 100 pages of version 1.7 of the PDF spec. I could easily enough write a script that downloads that from somewhere. I use it because it's a very complex document with just about everything in it and turns out to be very useful for benchmarking.

**Comment by m-holger** [comment]
I use the whole of the 1.7 spec. If in doubt, the download link can be found at https://pikepdf.readthedocs.io/en/latest/references/resources.html

**Comment by jberkenbilt** [comment]
Try `qpdf pdf17-spec.pdf -- pages . 1-100 -- out.pdf` and use the result. Using just the first 100 pages will dramatically improve your running time without sacrificing the value. Compared to the first cmake build, just after 10.6.3, I'm seeing about a 20% to 25% improvement in performance. I think that's the biggest boost qpdf has ever seen. That includes a change to QTC that I'm about to push up after it goes through CI. I think the PointerHolder -> shared_ptr change was a big contributor, but several of your tactical fixes help considerably as well. Hopefully when this lands, it will result in further improvements if we do it right.

**Comment by m-holger** [comment]
I rebased the branch and I think I took care of all your comments.

**Comment by m-holger** [comment]
#### Summary of commits

[Updated summary to reflect the recent commits. Unchanged content from the previous summary are marked as quotes]

The initial commits are geared towards demonstrating that splitting `QPDFObject` allows `QPDF::objectChanged` to be removed at an acceptable cost. I believe this has been achieved. 

A second batch of commits leverage the split of `QPDFObject` to reduce the footprint of `QPDFObjectHandle`. `QPDFObjectHandle::obj` is now the only data member of `QPDFObjectHandle`.

At the current time, the combined effect of the changes is an improvement in runtime of around 1%. However, the last dozen commits have not been tuned and there are a number of obvious areas for possible optimization .

#### Removal of  `QPDF::objectChanged` 

> The removal of `QPDF::objectChanged` and the related `m::ever_replaced_objects` is implemented in commit "Remove QPDF::ObjectChanged". The removal is made possible by mutating `obj_cache[og].object` in place (rather than replacing the existing object in the cache) in `QPDF::replaceObject` and `QPDF::swapObjects`.
> 
> The ability to mutate objects in place relies on the methods `QPDFObject::assign` and `QPDFObject::swapWith` introduced in commit "Split QPDFObject into QPDFObject and QPDFValue". These two methods in turn depend on the pdf object type being able to be changed by replacing the `QPDFObject::value` member rather than the entire `QPDFObject` itself.
> 
> The ability to mutate objects associated with a pdf file transparently without the need to refresh any object handles also relies on all object handles associated with an `QPDF` object to have a valid `QPDFObject` in their `obj` member. Commits "Add class QPDF_Unresolved" and "Modify newIndirect to set QPDFObjectHandle::obj" are in furtherance of this need, and commits "Remove QPDFObjectHandle::reserved" and "Remove QPDFObjectHandle::makeReserved" are associated clean-up commits. [From a simplicity perspective I considered to also implement uninitialized `QPDFObjectHandle`s as `QPDFObject`s rather than a bool flag, removing the need to ever check `QPDFObjectHandle::obj` for being a nullptr. However, this turned out to be ridiculously expensive.]
> 
> Commits "Replace calls to QPDFObjectHandle::Factory::newIndirect where possible" and "Add private methods QPDFObjectHandle::asArray etc" primary purpose is to facilitate the implementation of "Split QPDFObject into QPDFObject and QPDFValue".
> 
> The purpose of "Change QPDFObjectHandle::shallowCopyInternal to copy scalars" is to ensure that no object can appear more than once in the `QPDF` `obj_cache`.
> 
> The (possibly secondary) purpose of commits "Add private methods QPDFObjectHandle::asArray etc", "Change QPDFObjectHandle::shallowCopyInternal to copy scalars" and "Remove virtual methods QPDFValue::getTypeCode and getTypeName" is to mitigate the cost of the extra indirection involved in the split of `QPDFObject`.

#### Removal of `QPDFObjectHandle` data members

Commits  "Remove QPDFObjectHandle::initialized" to "Remove methods and parameters obsoleted by the last two commits" are in support of this aim.

`QPDFObjectHandle::reserved` was removed as part of the first batch of commits in "Remove QPDFObjectHandle::reserved".

Removal of `QPDFObjectHandle::initialized` is trivial given that an  `QPDFObjectHandle::obj` is null if and only if the the handle is uninitialized. This is implemented in commit  "Remove QPDFObjectHandle::initialized".

Removal of the `qpdf` and `og` data members can be broken down into four steps:

- Add   `qpdf` and `og` members to `QPDFValue`.

  This is one of the purposes of commit "Change QPDF_Unresolved::create method to take QPDF* and QPDFObjGen".

- Pass responsibility for maintaining the  `qpdf` and `og` members to `QPDF`.

  The main commits dealing with this step are  "Change QPDF_Unresolved::create method to take QPDF* and QPDFObjGen" and "Add methods QPDFObject::setObjGen and QPDFObject::resetObjGen". 

  Preparatory work for the above commits are in "Add private methods QPDF::isCached and QPDF::isUnresolved" and "Add private method QPDF::ObjCache::update".

- Ensure object resolution is transparent to object handles.

  We need to ensure that object resolution does not change individual  `QPDFObjectHandle` objects, but only the `QPDFObject` instances they point to. Commits "Modify QPDF::getObject to not to resolve the object" and "Change return type of QPDF::resolve to void" support this step, with tidy up in "Remove QPDFObjectHandle::newIndirect".

- Actually remove  `qpdf` and `og`.

  Commits "Add method QPDFObject::getQPDF and remove QPDFObjectHandle::qpdf" and "Add method QPDFObject::getQPDF and remove QPDFObjectHandle::og" deal with this, with tidy up in "Remove methods and parameters obsoleted by the last two commits".

**Comment by m-holger** [comment]
The last commit addresses the point @jbarlow83 makes in his comments in #747.

**Comment by jberkenbilt** [comment]
@m-holger Could you resolve the merge conflicts here? This is next on my to-do list. :-)

**Comment by jberkenbilt** [comment]
@m-holger Before you do anything with this, I'm working on it now and intend to merge some of the earlier commits. I'll let you know when I'm ready for you to resolve the merge conflict.

**Comment by jberkenbilt** [comment]
@m-holger I'm ready for you to rebase this to current main. If you do, would you mind using my object-value branch as a starting point? The tip of jberkenbilt/object-value is identical to your tidy3 branch. As I was working to understand the commits, I did a rebase -i and did minor cleanup, basically resolving my earlier PR comments by removing the offending lines at the time of introduction rather than having them cleaned up later. I think it was only the moving of the ot_unresolved declaration and the stray `;`. Anyway, there are no meaningful changes on my branch, but if you start there for your rebase, it will make it just a little easier for me to work on the changes. I'm going to pause my review until you've had a chance to rebase to current main, integration your changes into the QPDFParser class.

**Comment by jberkenbilt** [comment]
If you'd rather not start with jberkenbilt/object-value, I will just reintroduce my changes as I review. That's fine if it's your preference.

**Comment by m-holger** [comment]
I'll rebase tomorrow morning.

Sorry for the headers in QPDFParser. I made a mental note to clean them up before completely forgetting about it.

**Comment by m-holger** [comment]
Rebased  jberkenbilt/object-value.

If accepted, this will take care of the stray include file in QPDFParser.cc

**Comment by jberkenbilt** [comment]
Thanks -- I already fixed the stray header unless you fixed another one I missed. :-) I'll be able to spend most of the day on this today.

**Comment by m-holger** [comment]
From my recollection I expected to see more stray headers - the only one that I spotted was QPDF_Array.

**Review by jberkenbilt — CHANGES_REQUESTED** [review]
I'm convinced that this is the right approach. I have only minor comments at this point and would be ready to accept this PR today if you fix the small handful of remaining issues. I've put deep thought into this, and I have to admit that this solution is better than anything I had come up with. The fact that it was so easy to clear the owning PDF on destruction is another hint that this is a good change.

**Comment by jberkenbilt** [comment]
> * QPDFValue is an implementation detail and should not be exposed to the public API.

Yes, I agree. Actually I always regretted having exposed QPDFObject.hh. I didn't actually check, but if we can avoid exposing QPDFObject and QPDFValue, that would be great, or if not, maybe we can avoid exposing QPDFValue.

> * QPDFObject and QPDFValue are tightly coupled and performance critical. In a sense, QPDFObject provides the internal interface to QPDFValue for the rest of qpdf. In the circumstances I think it is reasonable for QPDFObject to directly access the data members of QPDFValue. Access to QPDFValues methods should be restricted to QPDFObject and QPDFValue's subclasses.

I agree, and I noticed that you were doing that. I think it's appropriate for the reasons you stated.

> * QPDFObject is an implementation detail and consideration should be given whether to schedule it for removal from the public API in qpdf 12. This would require a way of exposing  the object_type_e enumeration. All internal access should be through QPDFObjectHandle or QPDF.

I responded to your earlier point before seeing this. Yes, that's right -- that's why it was exposed, and I agree that we should remove it.

> Personally, I would be quite happy to make QPDFObjectHandle and QPDF friends of QPDFObject and do away with the ObjAccessor class.

Yeah, that's fine. Since QPDFObject is really not intended for public use, I agree that perhaps QPDFObjAccessor is overkill.

**Comment by jberkenbilt** [comment]
`qpdf_object_type_e` has been available since 10.5.0 which was released almost two years ago. I would be willing to break this for 11. I'd suggest merging what you have first, then we can do a separate change that changes the enumerated values of `qpdf_object_type_e` from `ot_...` to `qpdf_ot_...`, update interfaces as needed, and then move QPDFObject.hh private.

**Comment by jberkenbilt** [comment]
I would still like to leave the numerical values the same (not change the order) for reasons already explained (possible usage from other languages where people may have been forced to duplicate the numerical values).

**Comment by jberkenbilt** [comment]
Another reason that QPDFObject was previously exposed is because you needed the actual definition of a class to make a PointerHolder to it. You don't to make a std::shared_ptr -- a forward declaration is sufficient. So with the PointerHolder transition, this seems like the right release to move QPDFObject private.

**Comment by jberkenbilt** [comment]
Sorry for the noise. Looking at this again, I believe it would be trivial to move QPDFObject private. Let me know if you intend to do it in this PR, in a separate PR, or if you'd like me to do it. I trust you to do it, but I'm also happy to do it. It would just take me a few minutes, but I doubt it would take you any longer...

I will have to step away for a few hours soon. No rush on any of this -- we can merge when you're comfortable and when the small handful of issues are resolved, and I'll defer to you whether you want to fix the visibility before or after merge. We just have to do it before release.

**Comment by jberkenbilt** [comment]
One more note -- let's not change `ot_...` to `qpdf_ot_...`. That's too disruptive, especially for people who already switched to `qpdf_object_type_e`. Sorry for the churn.

**Comment by jberkenbilt** [comment]
Thanks for the fixes. I consider this to be mergeable. Let me know if you'd like me to merge it as is, and we can resume discussion of remaining issues after merge or if you'd like to do more on this before we merge. I'll wait until CI of course. It will be a few hours before I can touch it. No worries if no response today.

**Comment by jberkenbilt** [comment]
(actually I'm around another 45 minutes, then gone for a few hours)

**Comment by m-holger** [comment]
Please check  that I didn't miss anything. I had to deal with lots of other stuff and I'm not good at multi-tasking.

If I didn't miss any of your comments and CI is clear, lets merge as is and deal with the rest later.

**Comment by jberkenbilt** [comment]
You didn't miss anything. I'll merge when CI is clean or when I'm back from my appointment.

Thank you so much for all this work and persistence. It took me a while to be convinced, but I really think this is a solid improvement.

**Review by jberkenbilt — APPROVED** [review]

**Pull request merged**
